File: homy/integrations/uptime_kuma/widgets/status/status.py
"""Uptime Kuma status widget."""
import logging

from flask import jsonify, request

from homy.cache import cache_key, should_bypass_cache, widget_cache
from homy.integration_widget_util import get_integration_for_widget, get_widget_config
from homy import integration_widget_fetch as iwf

logger = logging.getLogger(__name__)

WIDGET = {
    'type': 'uptime_kuma_status',
    'name': 'Uptime Kuma Status',
    'integration_types': ['uptime_kuma'],
    'default_size_x': 8,
    'default_size_y': 5,
    'icon': 'activity',
    'config_schema': {
        'integration_id': {'type': 'text', 'label': 'Integration ID', 'default': ''},
        'view_mode': {
            'type': 'select',
            'label': 'Ansicht',
            'options': ['Liste', 'Kompakt'],
            'default': 'Liste',
        },
        'max_items': {
            'type': 'select',
            'label': 'Max. Monitore',
            'options': ['5', '10', '15', '20'],
            'default': '10',
        },
        'show_only_down': {
            'type': 'select',
            'label': 'Nur DOWN-Monitore',
            'options': ['Nein', 'Ja'],
            'default': 'Nein',
        },
        'show_status_badge': {
            'type': 'select',
            'label': 'Status-Badge anzeigen',
            'options': ['Ja', 'Nein'],
            'default': 'Ja',
        },
    },
}

CACHE_NS = 'uptime_kuma_status'
CACHE_TTL = 45


def register(app):
    @app.route('/api/uptime_kuma_status/data', methods=['GET'], endpoint='widget_uptime_kuma_status_data')
    def uptime_kuma_status_data():
        widget_id = request.args.get('widget_id')
        if not widget_id:
            return jsonify({'error': 'Bad Request', 'message': 'widget_id required'}), 400

        wcfg = get_widget_config(widget_id)
        if wcfg is None:
            return jsonify({'error': 'Not Found'}), 404

        integration, config, err, code = get_integration_for_widget(widget_id, ['uptime_kuma'])
        if err:
            return jsonify({'configured': False, 'online': False, 'message': err}), code

        ck = cache_key(CACHE_NS, widget_id)
        if not should_bypass_cache():
            cached = widget_cache.get(ck)
            if cached is not None:
                return jsonify(cached)

        try:
            result = iwf.fetch_uptime_kuma_monitors(config)
            monitors = result.get('monitors', [])
            if wcfg.get('show_only_down') == 'Ja':
                monitors = [m for m in monitors if m.get('status') == 'down']
            try:
                max_items = max(1, min(int(wcfg.get('max_items', 10) or 10), 30))
            except (TypeError, ValueError):
                max_items = 10
            result['monitors'] = monitors[:max_items]
            result['view_mode'] = wcfg.get('view_mode', 'Liste')
            result['show_status_badge'] = wcfg.get('show_status_badge', 'Ja') == 'Ja'
            result['configured'] = True
            result['online'] = True
            widget_cache.set(ck, result, ttl=CACHE_TTL)
            return jsonify(result)
        except Exception as exc:
            logger.warning('uptime_kuma_status failed: %s', exc)
            return jsonify({'configured': True, 'online': False, 'message': str(exc)}), 200

File: homy/integrations/uptime_kuma/widgets/status/status.js
window.WidgetRegistry.register('uptime_kuma_status', {
    async render(container, widgetData, config) {
        container.innerHTML = '<div class="widget-loading"><div class="spinner"></div></div>';
        try {
            const data = await API.request(`/api/uptime_kuma_status/data?widget_id=${widgetData.id}`);
            container.innerHTML = '';
            if (!data.online) {
                container.innerHTML = `<div class="widget-error text-center" style="padding:12px;">${data.message || 'Not available'}</div>`;
                return;
            }
            const t = (k, fb) => (window.i18n?.translate(k) !== k ? window.i18n.translate(k) : null) || fb;
            const monitors = data.monitors || [];
            const allOk = (data.down_count ?? 0) === 0;

            const badgeHtml = data.show_status_badge
                ? (allOk
                    ? `<div class="uk-global-badge uk-ok"><i data-lucide="check-circle"></i><span>${t('uptime_all_up', 'Alle online')}</span></div>`
                    : `<div class="uk-global-badge uk-issues"><i data-lucide="alert-triangle"></i><span>${data.down_count} DOWN</span></div>`)
                : '';

            const statsHtml = `<div class="uk-stats">
                <div class="uk-stat"><span class="uk-stat-val uk-color-up">${data.up_count ?? 0}</span><span class="uk-stat-label">UP</span></div>
                <div class="uk-stat"><span class="uk-stat-val uk-color-down">${data.down_count ?? 0}</span><span class="uk-stat-label">DOWN</span></div>
                <div class="uk-stat"><span class="uk-stat-val">${data.monitor_count ?? 0}</span><span class="uk-stat-label">${t('uptime_total', 'Total')}</span></div>
                ${badgeHtml}
            </div>`;

            let listHtml;
            if (data.view_mode === 'Kompakt') {
                listHtml = monitors.length
                    ? `<div class="uk-compact-grid">${monitors.map(m =>
                        `<div class="uk-compact-item ${m.status}" title="${m.name}">
                            <span class="uk-compact-dot"></span>
                            <span class="uk-compact-name">${m.name}</span>
                        </div>`).join('')}</div>`
                    : `<div class="uk-empty">${t('uptime_all_up', 'Alle Monitore online')}</div>`;
            } else {
                listHtml = `<div class="uk-list">${
                    monitors.length
                        ? monitors.map(m => `
                            <div class="uk-row ${m.status}">
                                <span class="uk-status-dot ${m.status}"></span>
                                <span class="uk-row-name">${m.name}</span>
                                <span class="uk-row-status ${m.status}">${m.status.toUpperCase()}</span>
                            </div>`).join('')
                        : `<div class="uk-empty"><i data-lucide="check-circle" class="uk-empty-icon"></i>${t('uptime_all_up', 'Alle Monitore online')}</div>`
                }</div>`;
            }

            container.innerHTML = `<div class="uk-container">${statsHtml}${listHtml}</div>`;
            window.refreshLucideIcons?.(container);
        } catch (err) {
            container.innerHTML = `<div class="widget-error"><i data-lucide="alert-triangle"></i><span>${err.message}</span></div>`;
            window.refreshLucideIcons?.(container);
        }
    },
});
